### PR TITLE
DATACASS-219 - On startup CREATE TABLE from entities should only add 'if not exists'.

### PR DESCRIPTION
**Pull request by jxblum**
Fixes [DATACASS-219](https://jira.spring.io/browse/DATACASS-219).

Note, I also fixed the issue with `SchemaAction.RECREATE`.

### CODE REVIEW

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/mapping/BasicCassandraMappingContext.java [hunk 1]
```diff
@@ -153,7 +153,7 @@ public void setApplicationContext(ApplicationContext applicationContext) throws
 
 	@Override
 	public boolean usesTable(TableMetadata table) {
-		return entitySetsByTableName.containsKey(table.getName());
+		return entitySetsByTableName.containsKey(cqlId(table.getName()));
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/mapping/BasicCassandraMappingContext.java]: 👍

File: spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cassandra.config;
+
+import static org.hamcrest.Matchers.equalTo;
```
Comment (mp911de) [inline spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java]: Not a big deal, but let's stick to the [Spring Data Formatter settings](https://github.com/spring-projects/spring-data-cassandra/blob/master/etc/eclipse-formatting.xml). Static imports are imported with using wildcards then.

bildschirmfoto 2016-05-11 um 09 17 28

File: spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.Session;
+
+/**
+ * The CassandraCqlSessionFactoryBeanUnitTests class is a test suite of test cases testing the contract
+ * and functionality of the {@link CassandraCqlSessionFactoryBean} class.
+ *
+ * @author John Blum
+ * @see org.junit.Rule
```
Comment (mp911de) [inline spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java]: How about omitting `@see` tags in tests?

File: spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+ * @see org.junit.Test
+ * @see org.junit.rules.ExpectedException
+ * @see org.junit.runner.RunWith
+ * @see org.mockito.Mock
+ * @see org.mockito.Mockito
+ * @see org.mockito.runners.MockitoJUnitRunner
+ * @see org.springframework.cassandra.config.CassandraCqlSessionFactoryBean
+ * @since 1.5.0
```
Comment (mp911de) [inline spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java]: Please add `@see DATACASS-xxx` ticket references to the tests

File: spring-cql/src/test/java/org/springframework/cassandra/core/util/CollectionUtilsUnitTests.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+		assertThat(array, is(notNullValue()));
+		assertThat(array.length, is(equalTo(0)));
+	}
+
+	void assertNonNullEmptyCollection(Collection<?> collection) {
+		assertThat(collection, is(notNullValue()));
+		assertThat(collection.isEmpty(), is(true));
+	}
+
+	<T> Iterable<T> newIterable(final T... elements) {
```
Comment (mp911de) [inline spring-cql/src/test/java/org/springframework/cassandra/core/util/CollectionUtilsUnitTests.java]: Please help me understand why `newIterable` is needed. Assumed `Arrays.asList(…)` or `new ArrayList<>(…)` was sufficient for the performed tests.

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBean.java [hunk 6]
```diff
@@ -13,109 +13,153 @@
...
+
+	private SchemaAction schemaAction = SchemaAction.NONE;
 
 	@Override
 	public void afterPropertiesSet() throws Exception {
-
 		super.afterPropertiesSet();
 
-		Assert.notNull(converter);
+		Assert.state(converter != null, "Converter must not be null");
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBean.java]: I guess there's no consistent rule for exceptions in `afterPropertiesSet` - even the core framework throws `IllegalArgument`/`NullPointer`/other exceptions. Was just wondering, all good here.

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBeanUnitTests.java [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+import com.datastax.driver.core.Metadata;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.TableMetadata;
+
+/**
+ * The CassandraSessionFactoryBeanUnitTests class is a test suite of test cases testing the contract and functionality
+ * of the {@link CassandraSessionFactoryBean} class.
+ *
+ * @author John Blum
+ * @see org.junit.Rule
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBeanUnitTests.java]: How about omitting `@see (classname)` tags in tests?

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBeanUnitTests.java [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+ * @see org.junit.Test
+ * @see org.junit.rules.ExpectedException
+ * @see org.junit.runner.RunWith
+ * @see org.mockito.Mock
+ * @see org.mockito.Mockito
+ * @see org.mockito.runners.MockitoJUnitRunner
+ * @see org.springframework.data.cassandra.config.CassandraSessionFactoryBean
+ * @since 1.5.0
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBeanUnitTests.java]: Please add `@see DATACASS-xxx` ticket references to the tests

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.TableMetadata;
+import com.datastax.driver.core.exceptions.AlreadyExistsException;
+
+/**
+ * The SchemaActionIntegrationTests class is a test suite of test cases testing the contract and behavior
+ * of various {@link SchemaAction}s on startup of a Spring configured, Cassandra application client.
+ *
+ * @author John Blum
+ * @see org.junit.Test
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java]: How about omitting `@see (classname)` tags in tests?

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+ * @see org.junit.Rule
+ * @see org.junit.rules.ExpectedException
+ * @see org.springframework.cassandra.test.integration.AbstractEmbeddedCassandraIntegrationTest
+ * @see org.springframework.cassandra.test.integration.KeyspaceRule
+ * @see org.springframework.data.cassandra.config.CassandraSessionFactoryBean
+ * @see org.springframework.data.cassandra.config.SchemaAction
+ * @see org.springframework.data.cassandra.config.java.AbstractCassandraConfiguration
+ * @since 1.0.0
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java]: `1.5.0`

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+ */
+public class SchemaActionIntegrationTests extends AbstractEmbeddedCassandraIntegrationTest {
```
Comment (mp911de) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java]: Please add `@see DATACASS-xxx` ticket references to the tests

**Comment by mp911de** [comment]
Like the direction where Spring Data Cassandra is heading to. Please apply [Spring Data Formatting](https://github.com/spring-projects/spring-data-cassandra/blob/master/etc/eclipse-formatting.xml) to the code. Can support you on setting it up with IntelliJ. Added some minor comments.

File: spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java (hunk 2 above)
Comment (jxblum) [inline spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java]: Will do.

Personally, I think it was a mistake for Java to ever allow wildcard imports.  It is always better to be explicit about what exactly is being imported (a dependency, used by a class), static or not.  It is a simple matter to include (auto) import classes from within an IDE so using wildcards has less value today then it did in the past.  Using wildcards also makes it more difficult, especially outside an IDE to determine where a conflict originates.  No where is this more apparent then with Mockito and Hamcrest Matchers.

File: spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java (hunk 3 above)
Comment (jxblum) [inline spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java]: Right; I suppose these are not all that useful in tests eh, ;-)

File: spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java (hunk 4 above)
Comment (jxblum) [inline spring-cql/src/test/java/org/springframework/cassandra/config/CassandraCqlSessionFactoryBeanUnitTests.java]: Will do.  Though, I do think we should be using the proper syntax for the `@see` tag when referencing external references.

File: spring-cql/src/test/java/org/springframework/cassandra/core/util/CollectionUtilsUnitTests.java (hunk 5 above)
Comment (jxblum) [inline spring-cql/src/test/java/org/springframework/cassandra/core/util/CollectionUtilsUnitTests.java]: Yeah not really needed.  Just wanted a pure `Iterable` object that was not based on some `Collection` (e.g. `List`) for true unit testing.  But it is of little consequence.

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBean.java (hunk 6 above)
Comment (jxblum) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBean.java]: I think the Exception is important in this case (well, in most cases) as it communicates either the state of the object/system (`IllegalStateException`), the validity of the values passed in (`IllegalArgumentException`) or whether the programmer (caller) made a mistake using the API (`NullPointerException`, a programmatical `Exception`).

In this, case the `IllegalStateException` represents that the object is not in a proper state to finish post initialization activities, therefore, the `IllegalStateException` is the appropriate `Exception`.

However, what is probably not appropriate is the message.  The `Exception` message should rather read...

> Converter was not properly initialized

Will change.

Side note...

I do think there is an excessive number of inappropriate `Assert.notNull(object)`, `Assert.isTrue(condition)`, etc in the Spring Data codebase that should be including a **message**.

One of the reasons for this is, the Spring `Assert` utility class uses a generic message when one is not given, for example... `[Assert.notNull(object)](https://github.com/spring-projects/spring-framework/blob/v4.2.6.RELEASE/spring-core/src/main/java/org/springframework/util/Assert.java#L126)`.

The problem is, if a constructor or method with multiple arguments (for [example](https://github.com/spring-projects/spring-data-commons/blob/master/src/main/java/org/springframework/data/mapping/model/AbstractPersistentProperty.java#L61-L62)) asserts not null on those arguments without a message, such as...

```
public Object someMethod(Object argument1, Object argument2, ...) {
    Assert.notNull(argument1);
    Assert.notNull(argument2);

    ...
}
```

Then, the caller has no idea which argument was null, given the user no context and helpful debug information.

Food for thought.

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBeanUnitTests.java (hunk 7 above)
Comment (jxblum) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBeanUnitTests.java]: Will do, thx.

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBeanUnitTests.java (hunk 8 above)
Comment (jxblum) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/config/CassandraSessionFactoryBeanUnitTests.java]: Will do.

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java (hunk 9 above)
Comment (jxblum) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java]: Will do.

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java (hunk 10 above)
Comment (jxblum) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java]: Good catch, thank you.

File: spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java (hunk 11 above)
Comment (jxblum) [inline spring-data-cassandra/src/test/java/org/springframework/data/cassandra/test/integration/config/SchemaActionIntegrationTests.java]: Will do.

**Comment by jxblum** [comment]
Added the [Spring Data Formatting](https://github.com/spring-projects/spring-data-cassandra/blob/master/etc/eclipse-formatting.xml) to my IDE for the Spring Data projects.  Thanks for the tip.

**Comment by jxblum** [comment]
Thank you @mp911de for the code review.  Good tips.  Cheers.

**Pull request merged**